File: internal/barcli/bm25.go
package barcli

// BM25 ranking for token lookup (ADR-0232).
//
// Adapted from nn/internal/note/bm25.go.
// Documents are grammar tokens; title = token name + label (weighted 5×),
// body = definition + heuristics + distinctions.

import (
	"math"
	"sort"
	"strings"

	"github.com/kljensen/snowball/english"
)

const (
	bm25K1      = 1.5
	bm25B       = 0.75
	titleWeight = 5
)

// tokenDoc is a BM25 document representing one grammar token.
type tokenDoc struct {
	id           string // "axis:token"
	title        string // name + " " + label
	body         string // definition + heuristics + distinctions + axis context (flat, for legacy callers)
	heuristics   string // heuristics text only
	distinctions string // distinctions text only
	definition   string // definition text only
}

// buildTokenDocs constructs the BM25 corpus from all grammar tokens.
// If axisFilter is non-empty, only tokens on that axis are included.
func buildTokenDocs(g *Grammar, axisFilter string) []tokenDoc {
	var docs []tokenDoc

	add := func(axis, token, label, definition string, heuristics, distinctions []string) {
		if axisFilter != "" && axis != axisFilter {
			return
		}
		heuristicsText := strings.Join(heuristics, " ")
		distinctionsText := strings.Join(distinctions, " ")
		body := definition
		if heuristicsText != "" {
			body += " " + heuristicsText
		}
		if distinctionsText != "" {
			body += " " + distinctionsText
		}
		if axisH := g.AxisLevelHeuristics(axis); len(axisH) > 0 {
			ah := strings.Join(axisH, " ")
			body += " " + ah
			heuristicsText += " " + ah
		}
		if axisDesc := g.AxisLevelDescription(axis); axisDesc != "" {
			body += " " + axisDesc
			heuristicsText += " " + axisDesc
		}
		if rc := g.AxisRoutingConcept(axis, token); rc != "" {
			body += " " + rc
			definition += " " + rc
		}
		docs = append(docs, tokenDoc{
			id:           axis + ":" + token,
			title:        token + " " + label,
			body:         body,
			heuristics:   strings.TrimSpace(heuristicsText),
			distinctions: strings.TrimSpace(distinctionsText),
			definition:   strings.TrimSpace(definition),
		})
	}

	if axisFilter == "" || axisFilter == "task" {
		for _, taskName := range g.GetAllTasks() {
			meta := g.Static.Metadata[taskName]
			dists := make([]string, 0, len(meta.Distinctions))
			for _, d := range meta.Distinctions {
				dists = append(dists, d.Token)
			}
			add("task", taskName, g.TaskLabel(taskName), g.TaskDescription(taskName), meta.Heuristics, dists)
		}
	}

	for axis, tokenMap := range g.Axes.Metadata {
		for tokenName, meta := range tokenMap {
			dists := make([]string, 0, len(meta.Distinctions))
			for _, d := range meta.Distinctions {
				dists = append(dists, d.Token)
			}
			add(axis, tokenName, g.AxisLabel(axis, tokenName), meta.Definition, meta.Heuristics, dists)
		}
	}

	for axis, tokenMap := range g.Persona.Metadata {
		for tokenName, meta := range tokenMap {
			slug := slugifyToken(tokenName)
			dists := make([]string, 0, len(meta.Distinctions))
			for _, d := range meta.Distinctions {
				dists = append(dists, d.Token)
			}
			add(axis, slug, g.PersonaLabel(axis, tokenName), meta.Definition, meta.Heuristics, dists)
		}
	}

	if axisFilter == "" {
		for _, p := range g.StarterPacks {
			body := p.Framing + " " + p.Command
			if len(p.Heuristics) > 0 {
				body += " " + strings.Join(p.Heuristics, " ")
			}
			docs = append(docs, tokenDoc{
				id:    "pack:" + p.Name,
				title: p.Name,
				body:  body,
			})
		}
		for name, seq := range g.Sequences {
			body := seq.Description + " " + seq.Example
			if len(seq.Heuristics) > 0 {
				body += " " + strings.Join(seq.Heuristics, " ")
			}
			docs = append(docs, tokenDoc{
				id:    "sequence:" + name,
				title: name,
				body:  body,
			})
		}
	}

	return docs
}

// bm25Tokenize splits text into lowercase tokens (len > 1).
func bm25Tokenize(s string) []string {
	s = strings.ToLower(s)
	var tokens []string
	for _, word := range strings.FieldsFunc(s, func(r rune) bool {
		return !('a' <= r && r <= 'z') && !('0' <= r && r <= '9')
	}) {
		if len(word) > 1 {
			tokens = append(tokens, english.Stem(word, false))
		}
	}
	return tokens
}

const rrfK = 60.0

// bm25ScoresRRF computes RRF-fused scores across four per-field corpora:
// title, heuristics, distinctions, definition. Each field is ranked independently
// by BM25; scores are fused as Σ 1/(k + rank_i(d)) with k=60.
// Returns a map from doc ID to RRF score; docs absent from all field rankings score 0 and are excluded.
func bm25ScoresRRF(docs []tokenDoc, query string) map[string]float64 {
	type fieldCorpus struct {
		docs []tokenDoc
	}
	corpora := []fieldCorpus{
		{docs: make([]tokenDoc, len(docs))}, // title
		{docs: make([]tokenDoc, len(docs))}, // heuristics
		{docs: make([]tokenDoc, len(docs))}, // distinctions
		{docs: make([]tokenDoc, len(docs))}, // definition
	}
	for i, d := range docs {
		corpora[0].docs[i] = tokenDoc{id: d.id, title: d.title}
		corpora[1].docs[i] = tokenDoc{id: d.id, title: d.heuristics}
		corpora[2].docs[i] = tokenDoc{id: d.id, title: d.distinctions}
		corpora[3].docs[i] = tokenDoc{id: d.id, title: d.definition}
	}

	rrf := make(map[string]float64)
	for _, corpus := range corpora {
		scores := bm25Scores(corpus.docs, query)
		if len(scores) == 0 {
			continue
		}
		type kv struct {
			id    string
			score float64
		}
		ranked := make([]kv, 0, len(scores))
		for id, s := range scores {
			ranked = append(ranked, kv{id, s})
		}
		sort.Slice(ranked, func(i, j int) bool { return ranked[i].score > ranked[j].score })
		for rank, entry := range ranked {
			rrf[entry.id] += 1.0 / (rrfK + float64(rank+1))
		}
	}
	return rrf
}

// bm25Scores computes BM25 relevance scores for each tokenDoc against the query.
// Returns a map from doc ID to score; docs scoring 0 are excluded.
func bm25Scores(docs []tokenDoc, query string) map[string]float64 {
	terms := bm25Tokenize(query)
	if len(terms) == 0 {
		return nil
	}

	type docInfo struct {
		tf  map[string]int
		len int
	}
	infos := make([]docInfo, len(docs))
	totalLen := 0
	for i, d := range docs {
		tf := make(map[string]int)
		for _, t := range bm25Tokenize(d.title) {
			tf[t] += titleWeight
		}
		for _, t := range bm25Tokenize(d.body) {
			tf[t]++
		}
		dlen := len(bm25Tokenize(d.title))*titleWeight + len(bm25Tokenize(d.body))
		infos[i] = docInfo{tf: tf, len: dlen}
		totalLen += dlen
	}

	N := float64(len(docs))
	avgdl := float64(totalLen) / math.Max(N, 1)

	idf := make(map[string]float64, len(terms))
	for _, term := range terms {
		df := 0
		for _, info := range infos {
			if info.tf[term] > 0 {
				df++
			}
		}
		idf[term] = math.Log((N-float64(df)+0.5)/(float64(df)+0.5) + 1)
	}

	scores := make(map[string]float64)
	for i, d := range docs {
		info := infos[i]
		score := 0.0
		for _, term := range terms {
			tf := float64(info.tf[term])
			if tf == 0 {
				continue
			}
			dl := float64(info.len)
			score += idf[term] * (tf * (bm25K1 + 1)) /
				(tf + bm25K1*(1-bm25B+bm25B*dl/avgdl))
		}
		if score > 0 {
			scores[d.id] = score
		}
	}
	return scores
}

File: web/src/lib/bm25.ts
// BM25 ranking for token filter (ADR-0232).
// Ported from internal/barcli/bm25.go.

import { stemmer } from 'stemmer';
import type { TokenMeta } from './grammar.js';

const BM25_K1 = 1.5;
const BM25_B = 0.75;
const TITLE_WEIGHT = 5;

export interface RankedToken {
	token: TokenMeta;
	score: number;
}

function tokenize(s: string): string[] {
	return s
		.toLowerCase()
		.split(/[^a-z0-9_]+/)
		.filter((w) => w.length > 1)
		.map(stemmer);
}

function docTitle(t: TokenMeta): string {
	return t.token + ' ' + t.label;
}

function docBody(t: TokenMeta): string {
	const parts: string[] = [t.description];
	if (t.metadata?.definition) parts.push(t.metadata.definition);
	if (t.metadata?.heuristics?.length) parts.push(t.metadata.heuristics.join(' '));
	if (t.metadata?.distinctions?.length)
		parts.push(t.metadata.distinctions.map((d) => d.token + ' ' + d.note).join(' '));
	return parts.join(' ');
}

// bm25Score returns a Map from token name to BM25 score.
// Tokens with score 0 are excluded from the map.
export function bm25Score(tokens: TokenMeta[], query: string): Map<string, number> {
	const terms = tokenize(query);
	if (terms.length === 0) return new Map();

	// Build per-doc term frequency maps.
	type DocInfo = { tf: Map<string, number>; len: number };
	const docs: DocInfo[] = tokens.map((t) => {
		const tf = new Map<string, number>();
		for (const term of tokenize(docTitle(t))) tf.set(term, (tf.get(term) ?? 0) + TITLE_WEIGHT);
		for (const term of tokenize(docBody(t))) tf.set(term, (tf.get(term) ?? 0) + 1);
		const len = tokenize(docTitle(t)).length * TITLE_WEIGHT + tokenize(docBody(t)).length;
		return { tf, len };
	});

	const N = docs.length;
	const totalLen = docs.reduce((s, d) => s + d.len, 0);
	const avgdl = totalLen / Math.max(N, 1);

	// IDF per term.
	const idf = new Map<string, number>();
	for (const term of terms) {
		const df = docs.filter((d) => (d.tf.get(term) ?? 0) > 0).length;
		idf.set(term, Math.log((N - df + 0.5) / (df + 0.5) + 1));
	}

	const scores = new Map<string, number>();
	tokens.forEach((t, i) => {
		const d = docs[i];
		let score = 0;
		for (const term of terms) {
			const tf = d.tf.get(term) ?? 0;
			if (tf === 0) continue;
			const dl = d.len;
			score +=
				(idf.get(term) ?? 0) *
				(tf * (BM25_K1 + 1)) /
				(tf + BM25_K1 * (1 - BM25_B + BM25_B * (dl / avgdl)));
		}
		if (score > 0) scores.set(t.token, score);
	});
	return scores;
}

const RRF_K = 60;

function cosineSimilarity(a: Float32Array, b: Float32Array): number {
	if (a.length !== b.length) return 0;
	let dot = 0;
	for (let i = 0; i < a.length; i++) dot += a[i] * b[i];
	return dot;
}

// Full text used to embed a token: description + heuristics + distinctions + routing_concept.
function tokenDocText(t: TokenMeta): string {
	const parts: string[] = [t.description];
	if (t.routing_concept) parts.push(t.routing_concept);
	if (t.metadata?.definition) parts.push(t.metadata.definition);
	if (t.metadata?.heuristics?.length) parts.push(t.metadata.heuristics.join(' '));
	if (t.metadata?.distinctions?.length)
		parts.push(t.metadata.distinctions.map((d) => d.token + ' ' + d.note).join(' '));
	return parts.join(' ');
}

// rrfFuse merges ranked lists via Reciprocal Rank Fusion: Σ 1/(k + rank_i(d)).
function rrfFuse(lists: string[][], k: number): Map<string, number> {
	const scores = new Map<string, number>();
	for (const list of lists) {
		list.forEach((id, idx) => {
			scores.set(id, (scores.get(id) ?? 0) + 1 / (k + idx + 1));
		});
	}
	return scores;
}

// Shared production cache: token name → embedding vector. Populated lazily on first hybrid search.
export const sharedTokenEmbCache = new Map<string, Float32Array>();

// hybridRankTokens fuses three ranked lists via RRF(k=60):
//   1. BM25-title (token name + label only)
//   2. BM25-body (definition + heuristics + distinctions only)
//   3. Cosine similarity (embeddings)
// embedder returns a unit-norm Float32Array or null to degrade to BM25-only.
// Token embeddings are cached in tokenEmbCache (pass a fresh Map in tests).
export async function hybridRankTokens(
	tokens: TokenMeta[],
	query: string,
	embedder: ((q: string) => Promise<Float32Array | null>) | null,
	tokenEmbCache: Map<string, Float32Array> = sharedTokenEmbCache
): Promise<RankedToken[]> {
	// BM25-title corpus: score using token+label as the entire document text.
	const titleOnlyTokens = tokens.map((t) => ({
		...t,
		description: t.token + ' ' + t.label,
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta));
	const titleScores = bm25Score(titleOnlyTokens, query);
	const titleList = [...titleScores.entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	// BM25-body corpus: score using definition+heuristics+distinctions only.
	const bodyText = (t: TokenMeta): string => {
		const parts: string[] = [];
		if (t.metadata?.definition) parts.push(t.metadata.definition);
		if (t.metadata?.heuristics?.length) parts.push(t.metadata.heuristics.join(' '));
		if (t.metadata?.distinctions?.length)
			parts.push(t.metadata.distinctions.map((d) => d.token + ' ' + d.note).join(' '));
		return parts.join(' ');
	};
	const bodyOnlyTokens = tokens.map((t) => ({
		...t,
		description: bodyText(t),
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta));
	const bodyScores = bm25Score(bodyOnlyTokens, query);
	const bodyList = [...bodyScores.entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	// Cosine similarity list.
	let queryVec: Float32Array | null = null;
	if (embedder) {
		try { queryVec = await embedder(query); } catch { queryVec = null; }
	}
	if (queryVec && embedder) {
		await Promise.all(tokens.map(async (t) => {
			if (tokenEmbCache.has(t.token)) return;
			const precomputed = (t.metadata as { embedding?: number[] } | null)?.embedding;
			if (precomputed) {
				tokenEmbCache.set(t.token, new Float32Array(precomputed));
				return;
			}
			try {
				const vec = await embedder(tokenDocText(t));
				if (vec) tokenEmbCache.set(t.token, vec);
			} catch { /* leave uncached */ }
		}));
	}
	const cosineList: string[] = queryVec
		? tokens
			.map((t) => {
				const vec = tokenEmbCache.get(t.token);
				return { id: t.token, score: vec ? Math.max(0, cosineSimilarity(queryVec!, vec)) : 0 };
			})
			.filter((x) => x.score > 0)
			.sort((a, b) => b.score - a.score)
			.map((x) => x.id)
		: [];

	// RRF fusion across all three lists.
	const fused = rrfFuse([titleList, bodyList, cosineList], RRF_K);
	const ranked: RankedToken[] = [];
	for (const t of tokens) {
		const score = fused.get(t.token) ?? 0;
		if (score > 0) ranked.push({ token: t, score });
	}
	ranked.sort((a, b) => b.score - a.score);
	return ranked;
}

// bm25RankTokens ranks tokens via RRF across four per-field corpora:
// title, heuristics, distinctions, definition. Tokens absent from all field rankings are excluded.
export function bm25RankTokens(tokens: TokenMeta[], query: string): RankedToken[] {
	const titleList = [...bm25Score(tokens.map((t) => ({
		...t,
		description: t.token + ' ' + t.label,
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta)), query).entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	const heuristicsList = [...bm25Score(tokens.map((t) => ({
		...t,
		description: t.metadata?.heuristics?.join(' ') ?? '',
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta)), query).entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	const distinctionsList = [...bm25Score(tokens.map((t) => ({
		...t,
		description: t.metadata?.distinctions?.map((d) => d.token + ' ' + d.note).join(' ') ?? '',
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta)), query).entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	const definitionList = [...bm25Score(tokens.map((t) => ({
		...t,
		description: t.metadata?.definition ?? '',
		metadata: { definition: '', heuristics: [], distinctions: [] }
	} as unknown as TokenMeta)), query).entries()].sort((a, b) => b[1] - a[1]).map(([id]) => id);

	const fused = rrfFuse([titleList, heuristicsList, distinctionsList, definitionList], RRF_K);
	const ranked: RankedToken[] = [];
	for (const t of tokens) {
		const score = fused.get(t.token) ?? 0;
		if (score > 0) ranked.push({ token: t, score });
	}
	ranked.sort((a, b) => b.score - a.score);
	return ranked;
}
